### DISCUSSION TITLE
[Bug]: using SPM within nipype

### DISCUSSION THREAD

**Discussion by scott-love** [neurodesk/neurodesk.github.io-hugo#400] [Bug Report]
### Version

1.4.0

### What platform are you seeing the problem on?

Local PC

### What happened

I am not 100% sure this is actually a Neurodesk question/problem/bug. 

I am trying to create a nipype workflow using a Jupyter notebook inside Neurodesk. The workflow requires several SPM interfaces (e.g., FieldMap). However, there seems to be a problem with calling the SPM functions; potentially a problem with the command used to call the SPM standalone version. I tried adding lmod.load('spm12/r7219') and the other available spm version. I also tried adding lmod.load('matlab/2022a') but I don't think that should be helpful as I don't have a licence. 

I found this [page](https://github.com/NeuroDesk/neurocontainers/tree/master/recipes/spm12), which lead me to try adding matlab_cmd = '/opt/spm12/run_spm12.sh /opt/mcr/v97/ script'
spm.SPMCommand.set_mlab_paths(matlab_cmd=matlab_cmd, use_mcr=True).
This didn't help and it doesn't look correct to me as I would expect a path to a container.

Anyway, I have tried a few things and so far failed to use an SPM node in nipype. All help would be greatly appreciated

Thanks,
Scott

### How to reproduce

Here is a simple example.
`
from nipype import Node, Workflow
from bids.layout import BIDSLayout
from nipype.interfaces.utility import IdentityInterface, Function
from nipype.interfaces.spm import FieldMap

calcvdmS1 = Node(FieldMap(echo_times = (4.92, 7.38),
                        blip_direction = 1,
                        total_readout_time = 54.15,
                        maskbrain = False),
                        use_mcr = True,
               name = "calcvdmS1")

calcvdmS1.inputs.epi_file = '/data/sub-03067/ses-1/func/sub-03067_ses-1_task-BIP_acq-CONTINU_run-1_bold.nii'
calcvdmS1.inputs.magnitude_file = '/data/sub-03067/ses-1/fmap/sub-03067_ses-1_magnitude1.nii'
calcvdmS1.inputs.phase_file = '/data/sub-03067/ses-1/fmap/sub-03067_ses-1_phasediff.nii'
calcvdmS1.inputs.use_mcr = True

calcvdmS1.run()
`

### Relevant log output or screenshot

```shell
240201-15:10:08,793 nipype.workflow INFO:
	 [Node] Setting-up "calcvdmS1" in "/tmp/tmp0hrqef24/calcvdmS1".
240201-15:10:08,809 nipype.workflow INFO:
	 [Node] Executing "calcvdmS1" <nipype.interfaces.spm.preprocess.FieldMap>
240201-15:10:08,841 nipype.workflow INFO:
	 [Node] Finished "calcvdmS1", elapsed time 0.001577s.
240201-15:10:08,842 nipype.workflow WARNING:
	 Storing result file without outputs
240201-15:10:08,843 nipype.workflow WARNING:
	 [Node] Error on "calcvdmS1" (/tmp/tmp0hrqef24/calcvdmS1)
---------------------------------------------------------------------------
NodeExecutionError                        Traceback (most recent call last)
Cell In[5], line 1
----> 1 calcvdmS1.run()

File /opt/conda/lib/python3.11/site-packages/nipype/pipeline/engine/nodes.py:527, in Node.run(self, updatehash)
    524 savepkl(op.join(outdir, "_inputs.pklz"), self.inputs.get_traitsfree())
    526 try:
--> 527     result = self._run_interface(execute=True)
    528 except Exception:
    529     logger.warning('[Node] Error on "%s" (%s)', self.fullname, outdir)

File /opt/conda/lib/python3.11/site-packages/nipype/pipeline/engine/nodes.py:645, in Node._run_interface(self, execute, updatehash)
    643     self._update_hash()
    644     return self._load_results()
--> 645 return self._run_command(execute)

File /opt/conda/lib/python3.11/site-packages/nipype/pipeline/engine/nodes.py:771, in Node._run_command(self, execute, copyfiles)
    769     # Always pass along the traceback
    770     msg += f"Traceback:\n{_tab(runtime.traceback)}"
--> 771     raise NodeExecutionError(msg)
    773 return result

NodeExecutionError: Exception raised while executing Node calcvdmS1.

Traceback:
	Traceback (most recent call last):
	  File "/opt/conda/lib/python3.11/site-packages/nipype/interfaces/base/core.py", line 397, in run
	    runtime = self._run_interface(runtime)
	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "/opt/conda/lib/python3.11/site-packages/nipype/interfaces/spm/base.py", line 386, in _run_interface
	    results = self.mlab.run()
	              ^^^^^^^^^^^^^^^
	  File "/opt/conda/lib/python3.11/site-packages/nipype/interfaces/base/core.py", line 397, in run
	    runtime = self._run_interface(runtime)
	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "/opt/conda/lib/python3.11/site-packages/nipype/interfaces/matlab.py", line 156, in _run_interface
	    runtime = super(MatlabCommand, self)._run_interface(runtime)
	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "/opt/conda/lib/python3.11/site-packages/nipype/interfaces/base/core.py", line 752, in _run_interface
	    raise IOError(
	OSError: No command "matlab" found on host d54e7c1b65a6. Please check that the corresponding package is installed.
```

**Comment by stebo85** [comment]
Dear @scott-love ,

Our problem is how nipype calls the compiled version of SPM - it assumes that SPM standalone is installed locally. There are two workaround for this:

Start the nipype-1.8.3 container from the application menu. In that container, we installed SPM standalone locally and now you can run things like:
```
import nipype.interfaces.spm as spm
from nipype.interfaces import fsl
from nipype.interfaces import afni

matlab_cmd = '/opt/spm12/run_spm12.sh /opt/mcr/v97/ script'
spm.SPMCommand.set_mlab_paths(matlab_cmd=matlab_cmd, use_mcr=True)

norm12 = spm.Normalize12()
norm12.inputs.image_to_align = '/neurodesktop-storage/sub-01_ses-01_7T_T1w_defaced.nii'
norm12.run()

btr = fsl.BET()
btr.inputs.in_file = '/neurodesktop-storage/sub-01_ses-01_7T_T1w_defaced.nii'
btr.inputs.frac = 0.4
btr.inputs.out_file = '/neurodesktop-storage/sub-01_ses-01_7T_T1w_defaced_brain.nii'
res = btr.run() 

edge3 = afni.Edge3()
edge3.inputs.in_file = '/neurodesktop-storage/sub-01_ses-01_7T_T1w_defaced.nii'
edge3.inputs.out_file = '/neurodesktop-storage/sub-01_ses-01_7T_T1w_defaced_edges.nii'
edge3.inputs.datum = 'byte'
res = edge3.run()
```


The alternative is to specify the absolute path to SPM standalone and then you can run this in the global jupyter notebook system of Neurodesk:
here is the example: https://github.com/NeuroDesk/example-notebooks/blob/main/books/structural_imaging/nipype_module_example.ipynb

```
import nipype.interfaces.spm as spm
from nipype.interfaces import fsl
from nipype.interfaces import afni

matlab_cmd = '/cvmfs/neurodesk.ardc.edu.au/containers/spm12_r7771_20230609/spm12_r7771_20230609.simg/opt/spm12/run_spm12.sh /cvmfs/neurodesk.ardc.edu.au/containers/spm12_r7771_20230609/spm12_r7771_20230609.simg/opt/mcr/v97/ script'
spm.SPMCommand.set_mlab_paths(matlab_cmd=matlab_cmd, use_mcr=True)

norm12 = spm.Normalize12()
norm12.inputs.image_to_align = './sub-01_ses-01_7T_T1w_defaced.nii'
norm12.run()
```

This workaround will trigger a warning "stty: 'standard input': Inappropriate ioctl for device", which you can ignore and we are hoping to find a more elegant way in the future.


Let us know if this works for you or if you have an idea how to make this work more elegantly!

Thank you
Steffen

**Reply by scott-love** [reply]
Hi @stebo85, 
Thanks for the help! 
I found the second option to be quite elegant :) Unfortunately, it doesn't work on my setup. I think the best way to describe it is that I used the nipype example you pointed me [towards](https://github.com/NeuroDesk/example-notebooks/blob/main/books/structural_imaging/nipype_module_example.ipynb) by running every cell (except the first 2 related to Google Colab) on ly system using the global Jupyter notebook system. Everything worked except for the second last cell that requires SPM, which gave the error below.

stty: 'standard input': Inappropriate ioctl for device
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /opt/conda/lib/python3.11/site-packages/nipype/utils/spm_docs.py:49, in _strip_header(doc)
     48 try:
---> 49     index = doc.index(hdr)
     50 except ValueError as e:

ValueError: substring not found

The above exception was the direct cause of the following exception:

OSError                                   Traceback (most recent call last)
Cell In[14], line 7
      4 from nipype.interfaces import afni
      6 matlab_cmd = '/cvmfs/neurodesk.ardc.edu.au/containers/spm12_r7771_20230609/spm12_r7771_20230609.simg/opt/spm12/run_spm12.sh /cvmfs/neurodesk.ardc.edu.au/containers/spm12_r7771_20230609/spm12_r7771_20230609.simg/opt/mcr/v97/ script'
----> 7 spm.SPMCommand.set_mlab_paths(matlab_cmd=matlab_cmd, use_mcr=True)
      9 norm12 = spm.Normalize12()
     10 norm12.inputs.image_to_align = './sub-01_ses-01_7T_T1w_defaced.nii'

File /opt/conda/lib/python3.11/site-packages/nipype/interfaces/spm/base.py:322, in SPMCommand.set_mlab_paths(cls, matlab_cmd, paths, use_mcr)
    320 cls._paths = paths
    321 cls._use_mcr = use_mcr
--> 322 info_dict = Info.getinfo(matlab_cmd=matlab_cmd, paths=paths, use_mcr=use_mcr)

File /opt/conda/lib/python3.11/site-packages/nipype/interfaces/spm/base.py:238, in Info.getinfo(klass, matlab_cmd, paths, use_mcr)
    235     klass._paths = paths
    236     return None
--> 238 out = sd._strip_header(out.runtime.stdout)
    239 out_dict = {}
    240 for part in out.split("|"):

File /opt/conda/lib/python3.11/site-packages/nipype/utils/spm_docs.py:51, in _strip_header(doc)
     49     index = doc.index(hdr)
     50 except ValueError as e:
---> 51     raise IOError("This docstring was not generated by Nipype!\n") from e
     53 index += len(hdr)
     54 index += 1

OSError: This docstring was not generated by Nipype!


I don't know what other info could be helpful here and I appreciate that if you cannot replicate it then it will be complicated to solve.

I will also try the first option but it was not my preferred.

Thanks,
Scott

**Reply by stebo85** [reply]
Can you reproduce the problem here: https://play.neurodesk.org/ ?

If not, can you give more details about your setup?

**Reply by scott-love** [reply]
I cannot reproduce the problem there - the SPM based cell runs; it does give the expected "stty: 'standard input': Inappropriate ioctl for device" warning but the process completes. 

I was also tried your first option above. I opened a Neurodesktop instance, selected the Neurodesk/workflows/nipype/code1.8.3 application and tested :

```
import nipype.interfaces.spm as spm

matlab_cmd = '/opt/spm12/run_spm12.sh /opt/mcr/v97/ script'
spm.SPMCommand.set_mlab_paths(matlab_cmd=matlab_cmd, use_mcr=True)

norm12 = spm.Normalize12()
norm12.inputs.image_to_align = '/data/code/sub-01_ses-01_7T_T1w_defaced.nii'
norm12.run()
```

This worked. So the problem does seem to be restrained to the global Jupyter notebook system (on my system).

My setup:
- Mac Studio
- Sonoma 14.2.1
- Apple M1 Max
- docker v 4.27.1 : resources 32 GB memory limit, 2 gb swap, 210 gb virtual disk
- In the context of trying to use containers for everything (hence why I am loving the idea of Neurodesk) I have installed almost no 'scientific computing stuff' on this machine.

Are there any other details that could help?

**Reply by scott-love** [reply]
I came across this neurostars [post](https://neurostars.org/t/valueerror-substring-not-found-when-running-spm-spmcommand-version-with-spm-standalone-and-mcr-installation/24082) and [https://miykael.github.io/nipype_tutorial/notebooks/advanced_spmmcr.html](https://miykael.github.io/nipype_tutorial/notebooks/advanced_spmmcr.html), which led me to try adding 

```
%env FORCE_SPMMCR=1
%env
```
into the nipype Jupyter example. This did not solve the problem but I wanted to note it here and to be sure that the %env code was an appropriate way to test this.

**Reply by stebo85** [reply]
Dear @scott-love - I can reproduce the problem on an Apple M1 Max! I am looking into it now.

**Reply by stebo85** [reply]
A quick update on this. I found that Rosetta is interfering with SPM standalone on Apple Silicon. I got SPM standalone to startup by deactivating Rosetta in Docker, but the error in Nipype is still there.

**Reply by stebo85** [reply]
Dear @scott-love - we identified that this is indeed a bug in Rosetta and Docker will hopefully ship an update soon. I keep you updated when it's resolved.

**Comment by scott-love** [comment]
Dear @stebo85, thanks for keeping me updated and for working on a solution.

**Reply by stebo85** [reply]
Dear @scott-love - In the latest MacOS (14.4) update in combination with the latest Docker (4.28.0 (139021)) update this problem is now fixed. Can you test on your end?

This is the minimal example that I successfully ran on my M1 Pro Max:
https://www.neurodesk.org/example-notebooks/structural_imaging/nipype_module_example.html

**Comment by scott-love** [comment]
Great news, I successfully ran the minimal example on an M3 Pro.
Thanks again for solving this!
